File: src/day3/main.jl
# read input
lines = transpose(parse.(Int, hcat(collect.(readlines("input.txt"))...)))

function most_common(M, j)::Int
	rows = size(M, 1)
	num_ones = sum(M[:, j])
	num_zeros = rows - num_ones
	return num_ones >= num_zeros ? 1 : 0
end

# takes the xor of most common array
least_common(M, j) = xor(most_common(M, j), 1)

# calculate gamma and epsilon
cols = size(lines, 2)
gamma = join(string(most_common(lines, j)) for j in 1:cols)
epsilon = join(string(least_common(lines, j)) for j in 1:cols)

# pt.1
@show parse(Int, gamma; base = 2) * parse(Int, epsilon; base = 2)

function slice_by_bit(M, j, bit)
	match_indicies = M[:, j] .== bit
	return M[match_indicies, :]
end

function find_factor(M, algo::Function)
	pos = 1
	while true
		bit = algo(M, pos)
		M = slice_by_bit(M, pos, bit)
		size(M, 1) == 1 && break
		pos += 1
	end
	result = join(string(x) for x in M)
	return result
end

# pt.2
oxygen = find_factor(lines, most_common)
carbon = find_factor(lines, least_common)

@show parse(Int, oxygen; base = 2) * parse(Int, carbon; base = 2)

File: src/day3/main.cpp
#include <iostream>
#include <algorithm>
#include <string>
#include <fstream>
#include <vector>
#include <unordered_map>

using namespace std;

template<typename T>
void outVec(vector<T> arr) {
	cout << "[ ";
	for (T a : arr) { cout << a << " "; }
	cout << "]" << endl; 
}

int mostCommon(vector<vector<int> > arr, int j) {
	int rows = arr.size();
	int num_ones = 0, num_zeros = 0;
	for (int i = 0; i < rows; i++) {
		if (arr[i][j] == 1) {
			num_ones++;
		} else if (arr[i][j] == 0) {
			num_zeros++;
		} else {
			cerr << "error not 1 or 0!" << endl;
			return -1;
		}
	}
	return num_ones >= num_zeros ? 1 : 0;
}

vector<int> leastCommon(vector<int> mostCommon) {
	vector<int> leastCommon;
	for (int i : mostCommon) {
		leastCommon.push_back(i^1);
	}

	return leastCommon;
}

int main() {
	// open file stream
	ifstream file ("input.txt");
	file.close(); // close file

	vector<vector<int>> arr = {{1, 0, 1, 1}, {0, 0, 1, 1}, {1, 1, 1, 1}};
	vector<int> apeArr = {1, 0, 0, 1};

	cout << (mostCommon(arr, 0)^1) << endl;
	outVec(leastCommon(apeArr));
}
